src/main/java/com/massivecraft/factions/zcore/frame/fupgrades/UpgradesListener.java

```diff
@@ -95,41 +99,207 @@ private void lowerSpawnerDelay(SpawnerSpawnEvent e, double multiplier) {
-        } else if (below.getType() == Material.CACTUS) {
-            Block above = e.getBlock().getLocation().add(0.0D, 1.0D, 0.0D).getBlock();
-            if (above.getType() == Material.AIR && above.getLocation().add(0.0D, -2.0D, 0.0D).getBlock().getType() != Material.AIR)
-                above.setType(Material.CACTUS);
+
+    Set<Material> mushroomGrowBlock = EnumSet.of(Material.MYCELIUM, Material.PODZOL, Material.CRIMSON_NYLIUM, Material.WARPED_NYLIUM);
+	
+    private boolean canSurvive(Block e){
+    	Material below = e.getLocation().add(0.0, -1.0, 0.0).getBlock().getType();
+    	Material origin= e.getType();
+    	byte lightLevel =e.getLightLevel();
+    	if (origin == Material.AIR) {	
+    		if(mushroomGrowBlock.contains(below))  {
+    		return true;
+    	}else if ((lightLevel < 13)  && (below.isSolid() == true)){
+    		return true;
+    	}
+      }return false;
+    }
+
+    Location spaceForShroom;
+	
+    private Location getSpaceForShroom(Location location) {
+    	for (int x = 1; x >= -1; x--) {
+    		for (int y = 1; y >= -1; y--) {
+    			for (int z = 1; z >= -1; z--) { 
+    				if (canSurvive(location.getBlock().getRelative(x, y, z)) == true) {
+    					this.spaceForShroom = location.getBlock().getRelative(x, y, z).getLocation();
+    				}
+    			}
+    		}
+    	}return spaceForShroom;
+    }	
+	
+    @EventHandler
+    public void onCropGrow(BlockGrowEvent e) {
+        FLocation floc = FLocation.wrap(e.getBlock().getLocation());
+        Faction factionAtLoc = Board.getInstance().getFactionAt(floc);
+        if (!factionAtLoc.isWilderness()) {
+            int level = factionAtLoc.getUpgrade("Crops");
+            int chance = FactionsPlugin.getInstance().getFileManager().getUpgrades().getConfig().getInt("fupgrades.MainMenu.Crops.Crop-Boost.level-" + level);
+            if (level == 0 || chance == 0) {
+            	return;
+            }
+            int randomNum = ThreadLocalRandom.current().nextInt(0, 100);
+            if (randomNum <= chance) { 	
+            	this.growCrop(e);
+            }
+          }
         }
+    
+    BlockGrowEvent cropGrowEvent;
+    Set<BlockFace> blockFaces = EnumSet.of(BlockFace.NORTH, BlockFace.SOUTH, BlockFace.EAST, BlockFace.WEST);
+    Set<Material> manualCrops = EnumSet.of(Material.WHEAT, Material.BEETROOTS, Material.CARROTS, Material.POTATOES, Material.NETHER_WART, Material.COCOA);;
+    Set<Material> autoCrops = EnumSet.of(Material.CACTUS, Material.SUGAR_CANE, Material.MELON, Material.PUMPKIN);;
+    Set<Material> autoAgeableCrops = EnumSet.of(Material.SUGAR_CANE, Material.CACTUS);
+    Set<Material> fruitCrops = EnumSet.of(Material.PUMPKIN, Material.MELON);
+    Block blockCrop;
+    Material typeCropNew;
+    BlockData dataCropNew;
+    Location locationCrop;
+    
+    private void growCrop(BlockGrowEvent e) {
+    	this.cropGrowEvent = e;
+    	this.blockCrop = cropGrowEvent.getNewState().getBlock();
```

Comment:
If these variabels are local to this function, don't use `this` `this` makes it a variable in the scope of the entire class (which is what `this` is, this class). If the variable doesn't need to persist outside of the function, you can just declare it as `blockCrop`.

If you are using it in another function, consider passing it to that function instead.